Adapt to simplified subsumption in GHC

Jurriaan Hage cryptically mentions [here](https://github.com/ghc-proposals/ghc-proposals/pull/287#issuecomment-574671874) that implementing the [simplified subsumption proposal](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0287-simplify-subsumption.rst) in GHC will break `singletons`. Now that I've had the chance to actually compile and test out the [corresponding MR](https://gitlab.haskell.org/ghc/ghc/merge_requests/2600) that implements this proposal, I can finally confirm that this is the case. Simplified subsumption will break code that looks like this:

```hs
sEx :: Sing x -> Sing y -> Sing (Const x y)
sEx sx sy = singFun2 @ConstSym0 sConst `applySing` sx `applySing` sy
```
```
Bug.hs:37:33: error:
    • Couldn't match type ‘Sing y0 -> Sing t’
                     with ‘forall (t1 :: b).
                           Sing t1 -> Sing (Apply (Apply ConstSym0 t) t1)’
      Expected type: Sing t -> SingFunction1 (Apply ConstSym0 t)
        Actual type: Sing t -> Sing y0 -> Sing (Const t y0)
    • In the second argument of ‘singFun2’, namely ‘sConst’
      In the first argument of ‘applySing’, namely
        ‘singFun2 @ConstSym0 sConst’
      In the first argument of ‘applySing’, namely
        ‘singFun2 @ConstSym0 sConst `applySing` sx’
    • Relevant bindings include
        sy :: Sing y (bound at Bug.hs:37:8)
        sx :: Sing x (bound at Bug.hs:37:5)
   |
37 | sEx sx sy = singFun2 @ConstSym0 sConst `applySing` sx `applySing` sy
   |                                 ^^^^^^
```

The reason is that the type of `singFun2`:

```hs
singFun2 :: forall f. SingFunction2 f -> Sing f
```

Is equivalent to the following:

```hs
singFun2 :: forall f. (forall t1. Sing t -> forall t2. Sing t2 -> Sing (f `Apply` t1 `Apply` t2)) -> Sing f
```

In other words, the only way that `singFun2 @ConstSym0 sConst` ever typechecked before was due to deep skolemization, which no longer occurs under simplified subsumption. A similar problem occurs for `singFun3`, `singFun4`, etc. as well.

This is a problem for `singletons` since its TH machinery singles all function applications to use `singFun{N}` + `applySing`, so singling any function applied to two or more arguments will fail to typecheck under simplified subsumption. I can envision two possible workarounds:

1. Change the definitions of `SingFunction2` _et al._ to put all the `forall`s up front. In other words, go from this:

   ```hs
   type SingFunction1 f = forall t. Sing t -> Sing (f `Apply` t)
   type SingFunction2 f = forall t. Sing t -> SingFunction1 (f `Apply` t)
   type SingFunction3 f = forall t. Sing t -> SingFunction2 (f `Apply` t)
   ...
   ```

   To this:

   ```hs
   type SingFunction1 f = forall t. Sing t -> Sing (f `Apply` t)
   type SingFunction2 f = forall t1 t2. Sing t1 -> Sing t2 -> Sing (f `Apply` t1 `Apply` t2)
   type SingFunction3 f = forall t1 t2 t3. Sing t1 -> Sing t2 -> Sing t3 -> Sing (f `Apply` t1 `Apply` t2 `Apply` t3)
   ...
   ```

   The upside to this workaround is that all call sites to `singFun{N}` will continue to work as they did before, with no additional changes required. The downside is that it would require making the definitions of `SingFunction{N}` much less concise than they used to be. (Note that we go up to `SingFunction8`!) We could regain a tiny bit of concision, however, through the use of an auxiliary type family:

   ```hs
   data HList :: [Type] -> Type where
     HNil :: HList '[]
     (:>) :: x -> HList xs -> HList (x:xs)
   infixr 5 :>
   
   type family Fun (args :: [Type]) (res :: Type) :: Type where
     Fun '[]        res = res
     Fun (arg:args) res = arg ~> Fun args res
   
   type family SingFunctionAux (f :: Fun args res) (ts :: HList args) :: Type where
     SingFunctionAux                f HNil      = Sing f
     SingFunctionAux @(_:args) @res f (t :> ts) = Sing t -> SingFunctionAux @args @res (f @@ t) ts
   
   type SingFunction1 f = forall t. SingFunctionAux f (t :> HNil)
   type SingFunction2 f = forall t1 t2. SingFunctionAux f (t1 :> t2 :> HNil)
   type SingFunction3 f = forall t1 t2 t3. SingFunctionAux f (t1 :> t2 :> t3 :> HNil)
   ...
   ```

   This is, admittedly, not much shorter. (If only type family equations let you [return polytypes on the right-hand side](https://gitlab.haskell.org/ghc/ghc/issues/9269), then this could be much shorter!)

2. Eta expand all call sites of `singFun2` _et al._ For example, this will typecheck, even under simplified subsumption:

   ```hs
   sEx :: Sing x -> Sing y -> Sing (Const x y)
   sEx sx sy = singFun2 @ConstSym0 (\s -> sConst s) `applySing` sx `applySing` sy
   ```

   In general, a call site to `singFun{N} @T f` (where `N` ≥ 2) will need to be eta expanded to `singFun{N} @T (\s_1 ... s_{N-1} -> f s_1 ... s_{N-1})`.

   The upside to this workaround is that many of `singFun{N}`'s call sites are generated with TH, so this workaround can be automated somewhat. The downside is that there is also code in the wild that invokes `singFun{N}` by hand, and all such code will have to be updated by hand as well.

Which workaround sounds better?

--- a/src/Data/Singletons/Internal.hs
+++ b/src/Data/Singletons/Internal.hs
@@ -436,37 +436,49 @@ singFun1 :: forall f. SingFunction1 f -> Sing f
 singFun1 f = SLambda f
 
 type SingFunction2 :: (a1 ~> a2 ~> b) -> Type
-type SingFunction2 f = forall t. Sing t -> SingFunction1 (f @@ t)
+type SingFunction2 f = forall t1 t2. Sing t1 -> Sing t2 -> Sing (f @@ t1 @@ t2)
 singFun2 :: forall f. SingFunction2 f -> Sing f
 singFun2 f = SLambda (\x -> singFun1 (f x))
 
 type SingFunction3 :: (a1 ~> a2 ~> a3 ~> b) -> Type
-type SingFunction3 f = forall t. Sing t -> SingFunction2 (f @@ t)
+type SingFunction3 f = forall t1 t2 t3.
+                       Sing t1 -> Sing t2 -> Sing t3
+                    -> Sing (f @@ t1 @@ t2 @@ t3)
 singFun3 :: forall f. SingFunction3 f -> Sing f
 singFun3 f = SLambda (\x -> singFun2 (f x))
 
 type SingFunction4 :: (a1 ~> a2 ~> a3 ~> a4 ~> b) -> Type
-type SingFunction4 f = forall t. Sing t -> SingFunction3 (f @@ t)
+type SingFunction4 f = forall t1 t2 t3 t4.
+                       Sing t1 -> Sing t2 -> Sing t3 -> Sing t4
+                    -> Sing (f @@ t1 @@ t2 @@ t3 @@ t4)
 singFun4 :: forall f. SingFunction4 f -> Sing f
 singFun4 f = SLambda (\x -> singFun3 (f x))
 
 type SingFunction5 :: (a1 ~> a2 ~> a3 ~> a4 ~> a5 ~> b) -> Type
-type SingFunction5 f = forall t. Sing t -> SingFunction4 (f @@ t)
+type SingFunction5 f = forall t1 t2 t3 t4 t5.
+                       Sing t1 -> Sing t2 -> Sing t3 -> Sing t4 -> Sing t5
+                    -> Sing (f @@ t1 @@ t2 @@ t3 @@ t4 @@ t5)
 singFun5 :: forall f. SingFunction5 f -> Sing f
 singFun5 f = SLambda (\x -> singFun4 (f x))
 
 type SingFunction6 :: (a1 ~> a2 ~> a3 ~> a4 ~> a5 ~> a6 ~> b) -> Type
-type SingFunction6 f = forall t. Sing t -> SingFunction5 (f @@ t)
+type SingFunction6 f = forall t1 t2 t3 t4 t5 t6.
+                       Sing t1 -> Sing t2 -> Sing t3 -> Sing t4 -> Sing t5 -> Sing t6
+                    -> Sing (f @@ t1 @@ t2 @@ t3 @@ t4 @@ t5 @@ t6)
 singFun6 :: forall f. SingFunction6 f -> Sing f
 singFun6 f = SLambda (\x -> singFun5 (f x))
 
 type SingFunction7 :: (a1 ~> a2 ~> a3 ~> a4 ~> a5 ~> a6 ~> a7 ~> b) -> Type
-type SingFunction7 f = forall t. Sing t -> SingFunction6 (f @@ t)
+type SingFunction7 f = forall t1 t2 t3 t4 t5 t6 t7.
+                       Sing t1 -> Sing t2 -> Sing t3 -> Sing t4 -> Sing t5 -> Sing t6 -> Sing t7
+                    -> Sing (f @@ t1 @@ t2 @@ t3 @@ t4 @@ t5 @@ t6 @@ t7)
 singFun7 :: forall f. SingFunction7 f -> Sing f
 singFun7 f = SLambda (\x -> singFun6 (f x))
 
 type SingFunction8 :: (a1 ~> a2 ~> a3 ~> a4 ~> a5 ~> a6 ~> a7 ~> a8 ~> b) -> Type
-type SingFunction8 f = forall t. Sing t -> SingFunction7 (f @@ t)
+type SingFunction8 f = forall t1 t2 t3 t4 t5 t6 t7 t8.
+                       Sing t1 -> Sing t2 -> Sing t3 -> Sing t4 -> Sing t5 -> Sing t6 -> Sing t7 -> Sing t8
+                    -> Sing (f @@ t1 @@ t2 @@ t3 @@ t4 @@ t5 @@ t6 @@ t7 @@ t8)
 singFun8 :: forall f. SingFunction8 f -> Sing f
 singFun8 f = SLambda (\x -> singFun7 (f x))
 

--- a/tests/ByHand.hs
+++ b/tests/ByHand.hs
@@ -861,8 +861,8 @@ sFindIndices :: forall a (t1 :: a ~> Bool) (t2 :: (List a)).
              -> Sing t2
              -> Sing (FindIndicesSym0 @@ t1 @@ t2)
 sFindIndices sP sLs =
-  let sLoop :: forall (u1 :: Nat). Sing u1
-            -> forall (u2 :: List a). Sing u2
+  let sLoop :: forall (u1 :: Nat) (u2 :: List a).
+               Sing u1 -> Sing u2
             -> Sing ((Let123LoopSym2 t1 t2) @@ u1 @@ u2)
       sLoop _ SNil = SNil
       sLoop sN (sX `SCons` sXs) = case sP @@ sX of

